### PR TITLE
Death Report

### PR DESCRIPTION
**Pull request by ranya-cfa**
@ekr

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -90,13 +94,23 @@ fn run_single_threaded(parameters_vec: Vec<Parameters>, output_path: &Path) {
...
+        );
         // Set up and execute context
         let mut context = Context::new();
         context.set_report_item_handler::<IncidenceReport>(move |item| {
             if let Err(e) = writer.serialize((Scenario { scenario }, item)) {
                 eprintln!("{}", e);
             }
         });
+        context.set_report_item_handler::<DeathReport>(move |item| {
+            if let Err(e) = death_writer.serialize((Scenario { scenario }, item)) {
```
Comment (ekr-cfa) [inline src/main.rs:136]: I see that this is in the original, but why is this error not fatal? @jasonasher ?

File: src/main.rs [hunk 2]
```diff
@@ -90,13 +94,23 @@ fn run_single_threaded(parameters_vec: Vec<Parameters>, output_path: &Path) {
                 .try_clone()
                 .expect("Could not write to output file"),
         );
+        let mut death_writer = writer_builder.from_writer(
```
Comment (ekr-cfa) [inline src/main.rs]: We want to rename `writer` here so that it's parallel with `death_writer`

File: src/main.rs [hunk 3]
```diff
@@ -106,10 +120,16 @@ fn run_single_threaded(parameters_vec: Vec<Parameters>, output_path: &Path) {
 fn run_multi_threaded(parameters_vec: Vec<Parameters>, output_path: &Path, threads: u8) {
     let output_file = File::create(output_path.join("incidence_report.csv"))
         .expect("Could not create output file.");
+    let death_file = File::create(output_path.join("death_report.csv"))
+        .expect("Could not create death report file.");
+
     let pool = ThreadPool::new(threads.into());
     let (sender, receiver) = channel();
+    let (death_sender, death_receiver) = channel();
```
Comment (ekr-cfa) [inline src/main.rs]: This implies we are going to have a new channel for each report type, which seems unwieldy, and at minimum is going to involve a lot of boilerplate. I tend to think what we want to do instead here is to have an enum that covers all our report types and then send that enum over the channel and demux in the subthread. @jasonasher wdyt?

File: src/sir/mod.rs [hunk 4]
```diff
@@ -6,3 +6,4 @@ pub mod person_properties;
 pub mod population_loader;
 pub mod transmission_manager;
 pub mod death_manager;
+pub mod death_report;
```
Comment (ekr-cfa) [inline src/sir/mod.rs]: Something wrong here with the file ending.

File: test/output/.Rhistory [hunk 5]
```diff
@@ -0,0 +1,46 @@
+library(tidyverse)
```
Comment (ekr-cfa) [inline test/output/.Rhistory]: This file looks like it was incorrectly added.

FYI, it's generally a good idea to go over your own PR to see if it has the contents you expect.

File: src/sir/death_report.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        })
+    }
+}
+
+impl Component for DeathReport {
+    fn init(context: &mut Context) {
+        context
+            .observe_person_property_changes::<DiseaseStatus>(handle_person_death);
+    }
+}
```
Comment (ekr-cfa) [inline src/sir/death_report.rs]: Please fix line ending

File: test/output/death_output.R
Comment (ekr-cfa) [inline test/output/death_output.R:1]: Can you please merge this with the other .R file to have a single script people run. Also, can you try to refactor the common pieces?

File: src/sir/death_report.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+) {
+    let disease_status = context.get_person_property_value::<DiseaseStatus>(person_id);
+    if matches!(disease_status, DiseaseStatus::D) {
+        context.release_report_item::<DeathReport>(Death {
+            time: context.get_time(),
+        })
+    }
+}
+
+impl Component for DeathReport {
```
Comment (ekr-cfa) [inline src/sir/death_report.rs:35]: Maybe it would make more sense to have just one Report component here. @jasonasher ?

**Review by ekr-cfa — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 1 above)
Comment (jasonasher) [reply src/main.rs:136]: Yes it should have just panicked in the original, too.

File: src/main.rs (hunk 3 above)
Comment (jasonasher) [reply src/main.rs]: My original idea was to have a channel for each report and absorb the boilerplate into the reports plugin (so you have a generic mechanism to add an arbitrary number of reports by type with associated output handlers, and then something that would collect all of the channels and pipe to output in the main thread). The enum approach could work too (and might be superior). I think both need some design iteration.

Comment (ekr-cfa) [reply src/main.rs]: @ranya-cfa do you want to try to prototype one or both of these?

File: src/main.rs (hunk 1 above)
Comment (ranya-cfa) [reply src/main.rs:136]: I just changed it to panic.

File: src/main.rs (hunk 3 above)
Comment (ranya-cfa) [reply src/main.rs]: I can start with the enum route today and see how that works.

File: src/main.rs [hunk 8]
```diff
@@ -102,14 +102,14 @@ fn run_single_threaded(parameters_vec: Vec<Parameters>, output_path: &Path) {
...
+        .serialize((Scenario { scenario }, item))
+        .unwrap_or_else(|e| panic!("{}", e));
+        }); 
         context.set_report_item_handler::<DeathReport>(move |item| {
-            if let Err(e) = death_writer.serialize((Scenario { scenario }, item)) {
-                eprintln!("{}", e);
-            }
+            death_writer
+        .serialize((Scenario { scenario }, item))
+        .unwrap_or_else(|e| panic!("{}", e));
```
Comment (ekr-cfa) [inline src/main.rs]: I think you could just use `unwrap()` here unless there is some predictable situation in which this would fail

File: test/output/plot_output.R [hunk 9]
```diff
@@ -1,11 +1,26 @@
...
     group_by(scenario, day = floor(time)) %>%
     summarize(incidence = n()) %>%
     mutate(scenario = factor(scenario))
 
-ggplot(summarized_incidence_report) +
-    geom_line(aes(x = day, y = incidence, color = scenario))
+summarized_death_report <- death_report %>%
+  group_by(scenario, day = floor(time)) %>%
+  summarize(death = n()) %>%
+  mutate(scenario = factor(scenario))
```
Comment (ekr-cfa) [inline test/output/plot_output.R]: Why not refactor this into a single function:

```
plot_report <- function(df, variable) {
   summarized <- df %>% group_by(scenario, day = floor(time)) %>%
   summarize(value = n()) %>%
   mutate(scenario = factor(scenario))

  plot <- ggplot(summarized) +
    geom_line(aes(x = day, y =value, color = scenario)) +
    ggtitle(paste(variable, " Report")

  ggsave(<make-name>, plot = plot)
}

plot_report(death_report, "death")
plot_report(incidence_report, "incidence")
```

**Review by ekr-cfa — CHANGES_REQUESTED** [review]

**Comment by ranya-cfa** [comment]
@ekr-cfa Is this ready to be merged? I just rebased to remove the enum change

File: src/main.rs [hunk 10]
```diff
@@ -120,17 +140,29 @@ fn run_multi_threaded(parameters_vec: Vec<Parameters>, output_path: &Path, threa
...
             context.execute();
             println!("Scenario {} completed", scenario);
         });
     }
     drop(sender);
+    drop(death_sender);
 
     // Write output from main thread
-    let mut writer = csv::Writer::from_writer(output_file);
+    let mut incidence_writer = csv::Writer::from_writer(output_file);
```
Comment (ekr-cfa) [inline src/main.rs:188]: As @jasonasher observed privately, this actually buffers all the death reports till all the incidence reports are done. Is there some way to listen to the next one on either channel?

**Review by ekr-cfa — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 10 above)
Comment (ekr-cfa) [reply src/main.rs:188]: Maybe using tokio select!

Comment (ranya-cfa) [reply src/main.rs:188]: Trying this now

File: src/main.rs [hunk 11]
```diff
@@ -76,6 +80,25 @@ fn setup_context(context: &mut Context, parameters: &Parameters) {
     context.add_component::<DeathManager>();
 }
 
+pub fn get_bounded_channel_report_handler<T: Report, S>(
```
Comment (k88hudson-cfa) [inline src/main.rs:84]: A comment about the intention to merge this into `eosim` proper would be good here, I think

File: src/main.rs [hunk 12]
```diff
@@ -103,38 +126,65 @@ fn run_single_threaded(parameters_vec: Vec<Parameters>, output_path: &Path) {
     }
 }
 
-fn run_multi_threaded(parameters_vec: Vec<Parameters>, output_path: &Path, threads: u8) {
+async fn run_multi_threaded(parameters_vec: Vec<Parameters>, output_path: &Path, threads: u8) {
     let output_file = File::create(output_path.join("incidence_report.csv"))
         .expect("Could not create output file.");
+    let death_file = File::create(output_path.join("death_report.csv"))
```
Comment (k88hudson-cfa) [inline src/main.rs:149]: I notice this is only implemented for `run_multi_threaded` – do you need to update `run_single_threaded` too? 

(Also: out of scope for this PR, but maybe we could figure out a way to DRY this out a bit?)

File: test/output/plot_output.R [hunk 13]
```diff
@@ -1,11 +1,22 @@
 library(tidyverse)
```
Comment (k88hudson-cfa) [inline test/output/plot_output.R:1]: I'm assuming you don't want to actually commit `Rplots.pdf`? Or was that intentional

**Review by k88hudson-cfa — CHANGES_REQUESTED** [review]
@ranya-cfa  just to clarify – IMO this looks like it works as expected / ready to merge to me with the two changes (removing the Rplots.pdf, adding death reports to the single threaded version).

Took me some time to wrap my head around the bounded channel stuff and file writing 😅. I think it makes more sense to iterate on that + the tests as it's merged into `eosim` v.s. doing more with it here

File: src/main.rs (hunk 12 above)
Comment (ranya-cfa) [reply src/main.rs:149]: I completely missed this on my end, apologies! I had it implemented in an earlier PR but must have lost it when I merged somewhere. Fixing that right now.

**Comment by ranya-cfa** [comment]
@k88hudson-cfa Thank you for taking a look at this! Does it look better now?

File: src/main.rs [hunk 14]
```diff
@@ -143,13 +210,96 @@ fn main() {
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tokio::sync::mpsc;
+    use tokio::time::{sleep, Duration};
+    use std::sync::Arc;
+    use tokio::sync::Mutex;
+    use futures::executor::block_on;
```
Comment (k88hudson-cfa) [inline src/main.rs]: This is unused, can you remove?

File: src/main.rs [hunk 15]
```diff
@@ -143,13 +210,96 @@ fn main() {
...
+            handler(4);
+            println!("Handler 4 sent");
+        });
+
+        // Now, let the producer block for a while
+        sleep(Duration::from_millis(500)).await;
+
+        // Check that the producer has not been able to send the third message yet
+        {
+            let mut flag = stop_flag.lock().await;
```
Comment (k88hudson-cfa) [inline src/main.rs]: Looks like this doesn't need to be mutable

**Review by k88hudson-cfa — CHANGES_REQUESTED** [review]
Yup looks great, thanks! I missed a couple of warnings in the test – can you fix those up and then let's squash and merge 🚢

**Comment by ranya-cfa** [comment]
@k88hudson-cfa Just finished cleaning up unused imports and the mut! Is it ready to merge?

**Review by k88hudson-cfa — APPROVED** [review]
Sweet! @jasonasher maybe you need to merge?
